### PR TITLE
Update conformer

### PR DESCRIPTION
**Pull request by VahidooX**
+ Polished the conformer code and configs.
+ Added mask_value to the specaug
+ Refactored positional embedding classes

### CODE REVIEW

File: nemo/collections/asr/parts/multi_head_attention.py [hunk 1]
```diff
@@ -211,40 +211,41 @@ def forward(self, query, key, value, mask, pos_emb):
...
                 return
         pe = torch.zeros(needed_size, self.d_model)
-        if self.reverse:
-            position = torch.arange(-(x.size(1) - 1), x.size(1), 1.0, dtype=torch.float32).unsqueeze(1)
-        else:
+        if self.pos_type == "Transformer":
             position = torch.arange(0, x.size(1), dtype=torch.float32).unsqueeze(1)
+        elif self.pos_type == "TransformerXL":
+            position = torch.arange(-(x.size(1) - 1), x.size(1), 1.0, dtype=torch.float32).unsqueeze(1)
+
```
Comment (titu1994) [inline nemo/collections/asr/parts/multi_head_attention.py:248]: Add else raise error case here.

File: nemo/collections/asr/parts/multi_head_attention.py [hunk 2]
```diff
@@ -211,40 +211,41 @@ def forward(self, query, key, value, mask, pos_emb):
...
 
     def extend_pe(self, x):
         """Reset the positional encodings."""
-        if self.reverse:
-            needed_size = 2 * x.size(1) - 1
-        else:
+        if self.pos_type == "Transformer":
             needed_size = x.size(1)
+        elif self.pos_type == "TransformerXL":
+            needed_size = 2 * x.size(1) - 1
```
Comment (titu1994) [inline nemo/collections/asr/parts/multi_head_attention.py:237]: Add else raise error case here.

File: nemo/collections/asr/parts/multi_head_attention.py [hunk 3]
```diff
@@ -211,40 +211,41 @@ def forward(self, query, key, value, mask, pos_emb):
...
     Args:
         d_model (int): embedding dim
         dropout_rate (float): dropout rate
         max_len (int): maximum input length
         reverse (int): whether to reverse the input position
     """
 
-    def __init__(self, d_model, dropout_rate, max_len=5000, reverse=False, xscale=None):
+    def __init__(self, d_model, dropout_rate, max_len=5000, xscale=None):
         """Construct an PositionalEncoding object."""
```
Comment (titu1994) [inline nemo/collections/asr/parts/multi_head_attention.py:223]: Constructor doesn't take pos_type argument ?

File: nemo/collections/asr/parts/conformer_modules.py [hunk 4]
```diff
@@ -95,7 +106,7 @@ def forward(self, x, att_mask=None, pos_emb=None, pad_mask=None):
 
         residual = x
         x = self.norm_self_att(x)
-        if self.self_attention_model == 'rel_pos':
+        if self.self_attention_model == 'rel_pos' or self.self_attention_model == 'rel_pos2':
```
Comment (titu1994) [inline nemo/collections/asr/parts/conformer_modules.py:98]: Document differences between rel pos types

File: nemo/collections/asr/models/ctc_models.py [hunk 5]
```diff
@@ -368,6 +371,9 @@ def output_types(self) -> Optional[Dict[str, NeuralType]]:
     def forward(
         self, input_signal=None, input_signal_length=None, processed_signal=None, processed_signal_length=None
     ):
+
+        # start = time.time()
+
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:376]: Remove timing code.

File: nemo/collections/asr/models/ctc_models.py [hunk 6]
```diff
@@ -384,16 +390,26 @@ def forward(
...
             processed_signal = self.spec_augmentation(input_spec=processed_signal)
 
+        # # REMOVE DEBUGGING
+        # import torch
+        # torch.manual_seed(0)
+        # processed_signal = torch.rand(processed_signal.shape).cuda()
+        # self.encoder(audio_signal=processed_signal[0:1,:,:400], length=processed_signal_length[0:1])[0][0:1,:,0:78]
+        # self.encoder(audio_signal=processed_signal[0:1,:,:500], length=processed_signal_length[0:1])[0][0:1,:,0:78]
+        # self.encoder(audio_signal=processed_signal[0:1,:,:], length=processed_signal_length[0:1])[0][0:1,:,0:78]
+
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:400]: Remove debugging

File: nemo/collections/asr/models/ctc_models.py [hunk 7]
```diff
@@ -384,16 +390,26 @@ def forward(
...
         encoded, encoded_len = self.encoder(audio_signal=processed_signal, length=processed_signal_length)
         log_probs = self.decoder(encoder_output=encoded)
         greedy_predictions = log_probs.argmax(dim=-1, keepdim=False)
+        # print("forward time:" + str(time.time() - start))
+
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:405]: Remove logging

File: nemo/collections/asr/models/ctc_models.py [hunk 8]
```diff
@@ -384,16 +390,26 @@ def forward(
...
         return log_probs, encoded_len, greedy_predictions
 
     # PTL-specific methods
     def training_step(self, batch, batch_nb):
         signal, signal_len, transcript, transcript_len = batch
         if isinstance(batch, DALIOutputs) and batch.has_processed_signal:
-            log_probs, encoded_len, predictions = self.forward(
+            log_probs, predictions_len, predictions = self.forward(
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:412]: This needs to be encoded_len same as else case

File: nemo/collections/asr/metrics/wer_bpe.py [hunk 9]
```diff
@@ -98,7 +100,9 @@ def ctc_decoder_predictions_tensor(self, predictions: torch.Tensor) -> List[str]
             hypotheses.append(hypothesis)
         return hypotheses
 
-    def update(self, predictions: torch.Tensor, targets: torch.Tensor, target_lengths: torch.Tensor):
+    def update(
+        self, predictions: torch.Tensor, targets: torch.Tensor, target_lengths: torch.Tensor, predictions_lengths=None
+    ):
```
Comment (titu1994) [inline nemo/collections/asr/metrics/wer_bpe.py:105]: Add type of prediction Len - torch.Tensor

File: nemo/collections/asr/metrics/wer.py [hunk 10]
```diff
@@ -135,7 +137,9 @@ def ctc_decoder_predictions_tensor(self, predictions: torch.Tensor) -> List[str]
             hypotheses.append(hypothesis)
         return hypotheses
 
-    def update(self, predictions: torch.Tensor, targets: torch.Tensor, target_lengths: torch.Tensor) -> torch.Tensor:
+    def update(
+        self, predictions: torch.Tensor, targets: torch.Tensor, target_lengths: torch.Tensor, predictions_lengths=None
+    ) -> torch.Tensor:
```
Comment (titu1994) [inline nemo/collections/asr/metrics/wer.py:142]: Add type to prediction_len - torch.Tensor

File: examples/asr/speech_to_text_rnnt_bpe.py [hunk 11]
```diff
@@ -75,6 +77,13 @@ def main(cfg):
     exp_manager(trainer, cfg.get("exp_manager", None))
     asr_model = EncDecRNNTBPEModel(cfg=cfg.model, trainer=trainer)
 
+    # initialize the model with the weights of the checkpoint specified by 'load_weights_from_checkpoint' in the configs
+    # You may use this option to start the training from a pre-trained checkpoint
+    checkpoint_path = cfg.model.get("load_weights_from_checkpoint", None)
```
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt_bpe.py:82]: Use with open_dict() and pop out this argument - it should not remain inside the config if submitted by user.

File: examples/asr/speech_to_text_rnnt_bpe.py [hunk 12]
```diff
@@ -75,6 +77,13 @@ def main(cfg):
...
+    if checkpoint_path:
+        logging.info(f'Initializing the model with the checkpoint at "{checkpoint_path}"')
+        asr_model.load_state_dict(torch.load(checkpoint_path, map_location=asr_model.device)["state_dict"])
+
```
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt_bpe.py:86]: Load checkpoint on seperate line, load weights and then explicitly del the entire checkpoint. Otherwise the optimizer state will stay over and won't get deleted before model creation + training begins.

File: examples/asr/speech_to_text_rnnt_bpe.py [hunk 13]
```diff
@@ -13,9 +13,11 @@
 # limitations under the License.
 
 import pytorch_lightning as pl
+import torch
```
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt_bpe.py:17]: Add from omegaconf import open_dict

File: examples/asr/speech_to_text_rnnt.py [hunk 14]
```diff
@@ -74,6 +90,13 @@ def main(cfg):
     exp_manager(trainer, cfg.get("exp_manager", None))
     asr_model = EncDecRNNTModel(cfg=cfg.model, trainer=trainer)
 
+    # initialize the model with the weights of the checkpoint specified by 'load_weights_from_checkpoint' in the configs
+    # You may use this option to start the training from a pre-trained checkpoint
+    checkpoint_path = cfg.model.get("load_weights_from_checkpoint", None)
```
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt.py:95]: Same comments as below - pop out this command line override, load checkpoint in seperate line, load checkpoint into model and explicitly del the entire checkpoint. 

X = torch.load()
Model.load_state_dict(X[state_dict])
del X

File: examples/asr/speech_to_text_rnnt.py [hunk 15]
```diff
@@ -13,13 +13,29 @@
 # limitations under the License.
 
 import pytorch_lightning as pl
+import torch
```
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt.py:16]: Add from omegaconf import open_dict

File: examples/asr/speech_to_text_bpe.py [hunk 16]
```diff
@@ -69,6 +70,13 @@ def main(cfg):
 
     asr_model = EncDecCTCModelBPE(cfg=cfg.model, trainer=trainer)
 
+    # initialize the model with the weights of the checkpoint specified by 'load_weights_from_checkpoint' in the configs
+    # You may use this option to start the training from a pre-trained checkpoint
+    checkpoint_path = cfg.model.get("load_weights_from_checkpoint", None)
+    if checkpoint_path:
```
Comment (titu1994) [inline examples/asr/speech_to_text_bpe.py:76]: Same comments as below

Comment (titu1994) [reply examples/asr/speech_to_text_bpe.py:76]: + add this even to the speech_to_text.py script please. Quite convenient

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks good, minor comments and some refactoring needed. @okuchaiev could you review update to WER calculation - it will be masked by sequence length now (if the model provides it)

Finally, let's drop all the vocab. It adds space to Nemo for no reason and we don't want dozens of vocab for every single dataset. The model trained on any dataset contains the tokekizer and the vocab inside it so it is unnecessary to add it as a data file. (we need to find some external place to store it - maybe as a GitHub release? @okuchaiev)

File: nemo/collections/asr/metrics/wer.py [hunk 17]
```diff
@@ -114,7 +114,7 @@ def __init__(
         self.add_state("scores", default=torch.tensor(0), dist_reduce_fx='sum', persistent=False)
         self.add_state("words", default=torch.tensor(0), dist_reduce_fx='sum', persistent=False)
 
-    def ctc_decoder_predictions_tensor(self, predictions: torch.Tensor) -> List[str]:
+    def ctc_decoder_predictions_tensor(self, predictions: torch.Tensor, predictions_len=None) -> List[str]:
```
Comment (titu1994) [inline nemo/collections/asr/metrics/wer.py:117]: Add type to prediction_len

File: nemo/collections/asr/metrics/wer_bpe.py [hunk 18]
```diff
@@ -77,7 +77,7 @@ def __init__(
         self.add_state("scores", default=torch.tensor(0), dist_reduce_fx='sum', persistent=False)
         self.add_state("words", default=torch.tensor(0), dist_reduce_fx='sum', persistent=False)
 
-    def ctc_decoder_predictions_tensor(self, predictions: torch.Tensor) -> List[str]:
+    def ctc_decoder_predictions_tensor(self, predictions: torch.Tensor, predictions_len=None) -> List[str]:
```
Comment (titu1994) [inline nemo/collections/asr/metrics/wer_bpe.py:80]: Add type to prediction_len

File: nemo/collections/asr/models/ctc_models.py [hunk 19]
```diff
@@ -185,7 +185,9 @@ def transcribe(self, paths2audio_files: List[str], batch_size: int = 4, logprobs
                         for idx in range(logits.shape[0]):
                             hypotheses.append(logits[idx][: logits_len[idx]])
                     else:
-                        hypotheses += self._wer.ctc_decoder_predictions_tensor(greedy_predictions)
+                        hypotheses += self._wer.ctc_decoder_predictions_tensor(
+                            greedy_predictions, predictions_len=logits_len
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:189]: @okuchaiev @vsl9 Please check if this is fine.

**Review by titu1994 — COMMENTED** [review]
Overall looks great, I'll let @okuchaiev  and @vsl9 review the change to WER before we merge

**Comment by okuchaiev** [comment]
@vsl9 , @jbalam-nv  and @blisc could you please double-check changes to WER computation: ``nemo/collections/asr/metrics/wer.py``

File: nemo/collections/asr/models/ctc_models.py (hunk 19 above)
Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py:189]: Do all models return logits_len?

Comment (VahidooX) [reply nemo/collections/asr/models/ctc_models.py:189]: Yes, all CTC_Models return it here : 
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/ctc_models.py#L390

It can be None and when it is None, it is ignored during WER calculation. But our current CTC models actually return the correct lengths.

For example, Jasper/QN return the lengths which are given by encoder here: 
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/modules/conv_asr.py#L192

Previously we did not use the lenghts return by the encoder at all in our codes. This would be the first time to actually use those.

Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py:189]: It seems fine then. Theoretically, the Jasper QuartzNet could make predictions past the current returned value due to the receptive field, but I don't think this makes a large practical difference

**Review by blisc — APPROVED** [review]
WER changes looks good to me

**Review by titu1994 — APPROVED** [review]
LGTM
